### PR TITLE
NP-3036 Separate the PublicationInstance building into a class

### PR DESCRIPTION
**Pull request by Eirik-Eide-Saur**

### CODE REVIEW

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinMapper.java [hunk 1]
```diff
@@ -385,16 +229,12 @@ private String extractMainTitle() {
                    .flatMap(Collection::stream);
     }
 
-    private CristinBookOrReportMetadata extractCristinBookReport() {
+    public CristinBookOrReportMetadata extractCristinBookReport() {
```
Comment (Eirik-Eide-Saur) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinMapper.java:232]: This and the method extractCristinJournalPublication method are made public as they need to be reach from the new class PublicationInstanceBuilder

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/PublicationInstanceBuilder.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    public static final String ERROR_PARSING_MAIN_CATEGORY = "Error parsing main category";
+    public static final String ERROR_PARSING_MAIN_OR_SECONDARY_CATEGORIES = "Error parsing main or secondary "
+            + "categories";
+
+    private final CristinObject cristinObject;
+    private final CristinMapper mapper;
+
+    public PublicationInstanceBuilder(CristinObject cristinObject) {
+        this.cristinObject = cristinObject;
+        this.mapper = new CristinMapper(cristinObject);
```
Comment (Eirik-Eide-Saur) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/PublicationInstanceBuilder.java:52]: Not a big fan of this solution. I do believe all the "extractData" methods should be moved into its own class, but for now this gives access to the methods needed both in this and the CristinMapper class.

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/PublicationInstanceBuilder.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+                .withPages(createMonographPages())
+                .build();
+    }
+
+    private PublicationInstance<? extends Pages> createChapterArticle() {
+        return new ChapterArticle.Builder().build();
+    }
+
+    private String extractPagesBegin() {
+        return mapper.extractCristinJournalPublication().getPagesBegin();
```
Comment (axthosarouris) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/PublicationInstanceBuilder.java:184]: A Mapper has a PublicationInstanceBuilder which has a Mapper for just getting a field from the CristinObect?

Comment (axthosarouris) [reply cristin-import/src/main/java/no/unit/nva/cristin/mapper/PublicationInstanceBuilder.java:184]: The only reason you are using this method is to handle the case where the cristinObject is null. Require that in the constructor and simplify your code

**Review by axthosarouris — CHANGES_REQUESTED** [review]

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/PublicationInstanceBuilder.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    public static final String ERROR_CRISTIN_OBJECT_IS_NULL = "CristinObject can not be null";
+    public static final String ERROR_PARSING_SECONDARY_CATEGORY = "Error parsing secondary category";
+    public static final String ERROR_PARSING_MAIN_CATEGORY = "Error parsing main category";
+    public static final String ERROR_PARSING_MAIN_OR_SECONDARY_CATEGORIES = "Error parsing main or secondary "
+            + "categories";
+
+    private final CristinObject cristinObject;
+
+    public PublicationInstanceBuilder(CristinObject cristinObject) {
+        Validate.notNull(cristinObject, ERROR_CRISTIN_OBJECT_IS_NULL);
```
Comment (axthosarouris) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/PublicationInstanceBuilder.java:52]: better use Objects.requireNonNull that is native Java. Also the class you are using is not supposed to be used by end-users of the library.

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/PublicationInstanceBuilder.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        this.cristinObject = cristinObject;
+    }
+
+    @SuppressWarnings("PMD.CognitiveComplexity")
```
Comment (axthosarouris) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/PublicationInstanceBuilder.java:56]: If quality check fail, it is an indication that things should be refactored even more, I have created a PR with a suggestion on breaking up this class even more.  Here is the link: https://github.com/BIBSYSDEV/nva-publication-api/pull/516 .

The suggestion is not finished, it has some replication in the exception handling but shows the general direction. If you like my suggestion you can merge my branch to yours and polish it where necessary.

File: cristin-import/src/test/java/no/unit/nva/cristin/lambda/CristinEntryEventConsumerTest.java [hunk 6]
```diff
@@ -261,7 +262,7 @@ public void handlerThrowsExceptionWhenSecondaryCategoryTypeIsNotKnown() throws J
 
         Throwable cause = exception.getCause();
         assertThat(cause, is(instanceOf(UnsupportedOperationException.class)));
-        assertThat(cause.getMessage(), is(containsString(CristinMapper.ERROR_PARSING_SECONDARY_CATEGORY)));
+        assertThat(cause.getMessage(), is(containsString(PublicationInstanceBuilder.ERROR_PARSING_SECONDARY_CATEGORY)));
```
Comment (axthosarouris) [inline cristin-import/src/test/java/no/unit/nva/cristin/lambda/CristinEntryEventConsumerTest.java:265]: "is" is not necessary here.

**Review by axthosarouris — CHANGES_REQUESTED** [review]
based on the comments above

**Review by axthosarouris — APPROVED** [review]

**Pull request merged**